// @yao-pkg/pkg/lib-es5/resolver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveModule = void 0;
const resolve_1 = require("resolve");
const resolve_exports_1 = require("resolve.exports");
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const common_1 = require("./common");
const log_1 = require("./log");
/**
 * Resolve using package.json "exports" field (ESM-style)
 * @param packageName - Package name (e.g., 'nanoid')
 * @param subpath - Subpath within package (e.g., './url-alphabet')
 * @param packageRoot - Absolute path to package root
 * @returns Resolved path or null if not found
 */
function resolveWithExports(packageName, subpath, packageRoot) {
    try {
        const packageJsonPath = path_1.default.join(packageRoot, 'package.json');
        if (!fs_1.default.existsSync(packageJsonPath)) {
            return null;
        }
        const pkg = JSON.parse(fs_1.default.readFileSync(packageJsonPath, 'utf8'));
        // Check if package has exports field
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const pkgAny = pkg;
        if (!pkgAny.exports) {
            return null;
        }
        // Use resolve.exports to handle the exports field
        // For pkg's context, we're bundling CJS code, so try 'require' condition first
        // Then fallback to 'import' for ESM-only packages
        let resolved = (0, resolve_exports_1.exports)(pkgAny, subpath, {
            require: true, // Try require first
        });
        // Fallback to import condition for ESM-only packages
        if (!resolved) {
            resolved = (0, resolve_exports_1.exports)(pkgAny, subpath, {
                require: false, // This enables import condition
            });
        }
        if (resolved) {
            // resolved can be a string or array
            const resolvedPath = Array.isArray(resolved) ? resolved[0] : resolved;
            const fullPath = path_1.default.join(packageRoot, resolvedPath);
            if (fs_1.default.existsSync(fullPath)) {
                return fullPath;
            }
        }
        return null;
    }
    catch (_error) {
        log_1.log.debug(`Failed to resolve with exports field: ${packageName}${subpath}`);
        return null;
    }
}
/**
 * Try to resolve a module specifier as an ESM package
 * @param specifier - Module specifier (e.g., 'nanoid', 'nanoid/url-alphabet')
 * @param basedir - Base directory for resolution
 * @returns Resolved path or null
 */
function tryResolveESM(specifier, basedir) {
    try {
        // Parse package name and subpath
        let packageName;
        let subpath;
        if (specifier.startsWith('@')) {
            // Scoped package: @org/pkg or @org/pkg/subpath
            const parts = specifier.split('/');
            packageName = `${parts[0]}/${parts[1]}`;
            subpath = parts.length > 2 ? `./${parts.slice(2).join('/')}` : '.';
        }
        else {
            // Regular package: pkg or pkg/subpath
            const slashIndex = specifier.indexOf('/');
            if (slashIndex === -1) {
                packageName = specifier;
                subpath = '.';
            }
            else {
                packageName = specifier.substring(0, slashIndex);
                subpath = `./${specifier.substring(slashIndex + 1)}`;
            }
        }
        // Use resolveSync to find the package root (works with pnpm and other layouts)
        // Try to resolve the package.json to get the package root
        try {
            const pkgJsonPath = (0, resolve_1.sync)(path_1.default.join(packageName, 'package.json'), {
                basedir,
                preserveSymlinks: false,
            });
            const packageRoot = path_1.default.dirname(pkgJsonPath);
            // Try to resolve with exports field
            const resolved = resolveWithExports(packageName, subpath, packageRoot);
            if (resolved) {
                return resolved;
            }
        }
        catch {
            // If package.json resolution fails, fall through to return null
        }
        return null;
    }
    catch {
        return null;
    }
}
/**
 * Resolve a module specifier with ESM support
 * Falls back to standard CommonJS resolution if ESM resolution fails
 *
 * @param specifier - Module specifier to resolve
 * @param options - Resolution options
 * @returns Resolved file path and ESM flag
 */
function resolveModule(specifier, options) {
    const { basedir, extensions = ['.js', '.json', '.node'] } = options;
    // First, try ESM-style resolution with exports field
    const esmResolved = tryResolveESM(specifier, basedir);
    if (esmResolved) {
        // Use isESMFile which walks up to find the correct package.json
        return {
            resolved: esmResolved,
            isESM: (0, common_1.isESMFile)(esmResolved),
        };
    }
    // Fallback to standard CommonJS resolution
    const resolved = (0, resolve_1.sync)(specifier, {
        basedir,
        extensions,
    });
    return {
        resolved,
        isESM: false, // CJS resolution
    };
}
exports.resolveModule = resolveModule;
