// @yao-pkg/pkg-fetch/lib-es5/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.system = exports.need = exports.getNodeVersion = exports.satisfyingNodeVersion = void 0;
const fs_1 = require("fs");
const promises_1 = require("fs/promises");
const path_1 = __importDefault(require("path"));
const semver_1 = __importDefault(require("semver"));
const expected_1 = require("./expected");
const system_1 = require("./system");
const system = __importStar(require("./system"));
exports.system = system;
const places_1 = require("./places");
const log_1 = require("./log");
const build_1 = __importDefault(require("./build"));
const utils_1 = require("./utils");
const patches_json_1 = __importDefault(require("../patches/patches.json"));
const package_json_1 = require("../package.json");
async function download({ tag, name }, local) {
    const url = `https://github.com/yao-pkg/pkg-fetch/releases/download/${tag}/${name}`;
    try {
        await (0, utils_1.downloadUrl)(url, local);
        await (0, utils_1.plusx)(local);
    }
    catch {
        return false;
    }
    return true;
}
async function exists(file) {
    try {
        await (0, promises_1.stat)(file);
        return true;
    }
    catch (error) {
        return false;
    }
}
function satisfyingNodeVersion(nodeRange) {
    const versions = Object.keys(patches_json_1.default)
        .filter((nv) => semver_1.default.satisfies(nv, nodeRange) || nodeRange === 'latest')
        .sort((nv1, nv2) => (semver_1.default.gt(nv1, nv2) ? 1 : -1));
    const nodeVersion = versions.pop();
    if (!nodeVersion) {
        throw (0, log_1.wasReported)(`No available node version satisfies '${nodeRange}'`);
    }
    return nodeVersion;
}
exports.satisfyingNodeVersion = satisfyingNodeVersion;
function getNodeVersion(nodeRange) {
    nodeRange = (0, system_1.abiToNodeRange)(nodeRange); // 'm48' -> 'node6'
    if (!(0, system_1.isValidNodeRange)(nodeRange)) {
        throw (0, log_1.wasReported)("nodeRange must start with 'node'");
    }
    if (nodeRange !== 'latest') {
        nodeRange = `v${nodeRange.slice(4)}`; // 'node6' -> 'v6' for semver
    }
    const nodeVersion = satisfyingNodeVersion(nodeRange);
    return nodeVersion;
}
exports.getNodeVersion = getNodeVersion;
async function need(opts) {
    // eslint-disable-line complexity
    const { forceFetch, forceBuild, dryRun, output, nodeRange } = opts || {};
    let { platform, arch } = opts || {};
    if (!nodeRange)
        throw (0, log_1.wasReported)('nodeRange not specified');
    if (!platform)
        throw (0, log_1.wasReported)('platform not specified');
    if (!arch)
        throw (0, log_1.wasReported)('arch not specified');
    platform = (0, system_1.toFancyPlatform)(platform); // win32 -> win
    arch = (0, system_1.toFancyArch)(arch); // ia32 -> x86
    const nodeVersion = getNodeVersion(nodeRange);
    const fetched = (0, places_1.localPlace)({
        from: 'fetched',
        arch,
        nodeVersion,
        platform,
        version: package_json_1.version,
        output,
    });
    const built = (0, places_1.localPlace)({
        from: 'built',
        arch,
        nodeVersion,
        platform,
        version: package_json_1.version,
        output,
    });
    const remote = (0, places_1.remotePlace)({ arch, nodeVersion, platform, version: package_json_1.version });
    let fetchFailed;
    if (!forceBuild) {
        if (await exists(fetched)) {
            if (dryRun) {
                return 'exists';
            }
            // when node path is set, skip hash check
            if (!!process.env.PKG_NODE_PATH ||
                (await (0, utils_1.hash)(fetched)) === expected_1.EXPECTED_HASHES[remote.name]) {
                return fetched;
            }
            log_1.log.info('Binary hash does NOT match. Re-fetching...');
            (0, fs_1.unlinkSync)(fetched);
        }
    }
    if (!forceFetch) {
        if (await exists(built)) {
            if (dryRun)
                return 'exists';
            if (forceBuild)
                log_1.log.info('Reusing base binaries built locally:', built);
            return built;
        }
    }
    if (!forceBuild) {
        if (dryRun)
            return 'fetched';
        if (await download(remote, fetched)) {
            if ((await (0, utils_1.hash)(fetched)) === expected_1.EXPECTED_HASHES[remote.name]) {
                return fetched;
            }
            (0, fs_1.unlinkSync)(fetched);
            throw (0, log_1.wasReported)('Binary hash does NOT match.');
        }
        fetchFailed = true;
    }
    if (!dryRun && fetchFailed) {
        log_1.log.info('Not found in remote cache:', JSON.stringify(remote));
        if (forceFetch) {
            throw (0, log_1.wasReported)(`Failed to fetch.`);
        }
    }
    if (!dryRun) {
        log_1.log.info('Building base binary from source:', path_1.default.basename(built));
    }
    if (system_1.hostPlatform !== platform) {
        if ((system_1.hostPlatform !== 'alpine' || platform !== 'linuxstatic') && arch !== 'ppc64') {
            throw (0, log_1.wasReported)(`Not able to build for '${opts.platform}' here, only for '${system_1.hostPlatform}'`);
        }
    }
    if (system_1.knownArchs.indexOf(arch) < 0) {
        throw (0, log_1.wasReported)(`Unknown arch '${opts.arch}'. Specify ${system_1.knownArchs.join(', ')}`);
    }
    if (dryRun) {
        return 'built';
    }
    await (0, build_1.default)(nodeVersion, arch, platform, built);
    return built;
}
exports.need = need;
