// @yao-pkg/pkg-fetch/lib-es5/places.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.remotePlace = exports.localPlace = exports.cachePath = void 0;
const semver_1 = require("semver");
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const { PKG_CACHE_PATH } = process.env;
const IGNORE_TAG = Boolean(process.env.PKG_IGNORE_TAG);
exports.cachePath = PKG_CACHE_PATH || path_1.default.join(os_1.default.homedir(), '.pkg-cache');
function tagFromVersion(version) {
    const mj = (0, semver_1.major)(version);
    const mn = (0, semver_1.minor)(version);
    return `v${mj}.${mn}`;
}
function localPlace({ from, output, version, nodeVersion, platform, arch, }) {
    let binDir;
    if (process.env.PKG_NODE_PATH) {
        return path_1.default.resolve(process.env.PKG_NODE_PATH);
    }
    if (output) {
        binDir = path_1.default.resolve(output);
    }
    else {
        binDir = IGNORE_TAG
            ? path_1.default.join(exports.cachePath)
            : path_1.default.join(exports.cachePath, tagFromVersion(version));
    }
    return path_1.default.resolve(binDir, `${output ? 'node' : from}-${nodeVersion}-${platform}-${arch}`);
}
exports.localPlace = localPlace;
function remotePlace({ version, nodeVersion, platform, arch, }) {
    return {
        tag: tagFromVersion(version),
        name: `node-${nodeVersion}-${platform}-${arch}`,
    };
}
exports.remotePlace = remotePlace;
